languageServer/src/TextDocumentWrapper.ts

```diff
@@ -0,0 +1,294 @@
+        },
+        label: 'Add stage'
+      }
+    ]
+    for (const path of this.pythonFilePaths) {
+      completionTemplates.push({
+        body: {
+          cmd: `python ${path} $1`
+        },
+        label: `cmd: ${path}`
+      })
+    }
+
+    return completionTemplates
+  }
+
+  private getTemplateExpressionSymbolsInsideScalar(
+    scalarValue: string,
+    nodeOffset: number
+  ) {
+    const templateSymbols: DocumentSymbol[] = []
+
+    const templates = scalarValue.matchAll(variableTemplates)
+    for (const template of templates) {
+      const expression = template[1]
+      const expressionOffset: number = nodeOffset + (template.index ?? 0) + 2 // To account for the '${'
+      const symbols = expression.matchAll(/[\dA-Za-z]+/g)
+      for (const templateSymbol of symbols) {
+        const symbolStart = (templateSymbol.index ?? 0) + expressionOffset
+        const symbolEnd = symbolStart + templateSymbol[0].length
+        const symbolRange = Range.create(
+          this.positionAt(symbolStart),
+          this.positionAt(symbolEnd)
+        )
+        templateSymbols.push(
+          DocumentSymbol.create(
+            templateSymbol[0],
+            undefined,
+            SymbolKind.Variable,
+            symbolRange,
+            symbolRange
+          )
+        )
+      }
+    }
+
+    return templateSymbols
+  }
+
+  private yamlScalarNodeToDocumentSymbols(
+    node: Scalar,
+    [nodeStart, valueEnd, nodeEnd]: [number, number, number]
+  ) {
+    const nodeValue = `${node.value}`
+
+    const symbolsSoFar: DocumentSymbol[] = [
+      DocumentSymbol.create(
+        nodeValue,
+        undefined,
+        SymbolKind.String,
+        Range.create(this.positionAt(nodeStart), this.positionAt(nodeEnd)),
+        Range.create(this.positionAt(nodeStart), this.positionAt(valueEnd))
+      ),
+      ...this.getTemplateExpressionSymbolsInsideScalar(nodeValue, nodeStart)
+    ]
+
+    return symbolsSoFar
+  }
+
+  private calculateTextEdits(newText: string) {
+    const currentText = this.getText()
+    const edit = TextEdit.replace(
+      Range.create(this.positionAt(0), this.positionAt(currentText.length)),
+      newText
+    )
+
+    return [edit]
+  }
+
+  private yamlNodeToDocumentSymbols(
+    node: Node | Pair,
+    range: [number, number, number]
+  ): DocumentSymbol[] {
+    if (isNode(node)) {
+      if (isScalar(node)) {
+        return this.yamlScalarNodeToDocumentSymbols(node, range)
+      }
+    } else if (isPair(node)) {
+      return this.yamlNodeToDocumentSymbols(node.value as Node | Pair, range)
+    }
+
+    return []
+  }
+
+  private symbolScopeAt(position: Position): DocumentSymbol[] {
+    const cursorOffset: number = this.offsetAt(position)
+
+    const symbolsFound: Array<DocumentSymbol | null> = []
+
+    visit(this.getYamlDocument(), (_key, node) => {
+      if (isNode(node)) {
+        const range = node.range
+        if (!!range && cursorOffset >= range[0] && cursorOffset <= range[2]) {
```

Comment:
Are you looking to see if the cursor is in the range here? Can we add a variable to explain what is going on?